### PR TITLE
Reduced network traffic

### PR DESCRIPTION
**Pull request by akash-07**
**Feature description:** Large cluster sets are explicitly requested and cached by the workers. The controller does not send them along with every partial merge request.

**Parameters:** The max set size parameter for the request to be considered as `LargePartialMerge` instead of plain `PartialMerge` can be set in the config file using the field `max_set_size`.

Default is 500 B (which means all requests go as `LargePartialMerge` requests).

### CODE REVIEW

**Comment by akash-07** [comment]
The pattern is REQ-REP only.

File: src/dist/controller.cc [hunk 1]
```diff
@@ -214,6 +235,51 @@ agd::Status Controller::Run(const Params& params,
...
+        auto partial_it = partial_merge_map_.find(id);
+        if (partial_it == partial_merge_map_.end()) {
+          cout << "the set with " << id << " is not in the map \n";
+          exit(0);
+        }
+        partial_item = &partial_it->second;
+        // cout << "Some data --> " << partial_item->buf.size() << std::endl;
+
+        // creating a copy since zmq takes ownership without creating a copy
+        agd::Buffer buf;
```
Comment (sbyma) [inline src/dist/controller.cc:270]: use the `Buffer(size_t sz)` constructor to set the correct buffer size at construction and avoid a reallocation.

File: src/dist/controller.h [hunk 2]
```diff
@@ -83,16 +90,20 @@ class Controller {
...
       num_expected = other.num_expected;
       num_received.store(other.num_received.load());
+      buf = std::move(other.buf);
       return *this;
     }
     PartialMergeSet partial_set;
     uint32_t num_expected;
     std::atomic_uint_fast32_t num_received;
+    // holds the MarshalledClusterSet responding to set requests
+    agd::Buffer buf;
```
Comment (sbyma) [inline src/dist/controller.h]: name it something more appropriate, like `marshalled_set_buf` or something like that.

File: src/dist/dist_cluster.cc [hunk 3]
```diff
@@ -348,8 +369,9 @@ int main(int argc, char* argv[]) {
     params.request_queue_port = request_queue_port;
     params.response_queue_port = response_queue_port;
     params.incomplete_request_queue_port = incomplete_request_queue_port;
+    params.set_request_port = set_request_port;
     signal(SIGUSR1, my_handler);
```
Comment (sbyma) [inline src/dist/dist_cluster.cc]: use a more appropriate name than "my_handler".

File: src/dist/worker.h [hunk 4]
```diff
@@ -62,6 +65,16 @@ class Worker {
   std::thread incomplete_request_queue_thread_;
   bool irqt_signal_ = false;
 
+  // cache for partial merge sets, buf is the buf of MarshalledClusterSet
+  std::unordered_map<int, agd::Buffer> set_map;
```
Comment (sbyma) [inline src/dist/worker.h]: private member variable should have trailing `_`  ==> `set_map_`

File: src/dist/worker.h [hunk 5]
```diff
@@ -72,4 +85,5 @@ class Worker {
   std::string response_queue_address;
```
Comment (sbyma) [inline src/dist/worker.h:85]: these address strings are only referenced in one place and probably can just be local vars and not private members?

**Review by sbyma — CHANGES_REQUESTED** [review]
Just a few minor things to clean up

**Pull request merged**